Documentation/proposals/202411-zone-aware-sharding.md

```diff
@@ -0,0 +1,335 @@
+      values:
+        - europe-west4-a
+        - europe-west4-b
+```
+
+And the relabel configuration being genereated through this template
+
+```yaml
+{{ $shardsPerZone := max 1 (.spec.shards | div (len .spec.shardingStrategy.topology.values)) }}
+{{ $promZoneIdx := .operator.shardIndex | mod $shardsPerZone }}
+
+{{ .spec.shardingStrategy.additionalRelabelConfig | toYaml }}
+- source_labels: [ '{{ default "__meta_kubernetes_pod_label_topology_kubernetes_io_zone" .spec.shardingStrategy.topology.sourceLabel }}' ]
+  separator: ;
+  regex: '{{ index .spec.shardingStrategy.topology $promZoneIdx }}'
+  replacement: '$1'
+  action: 'keep'
+- source_labels: [ '__address__' ] 
+  separator: ;
+  modulus: {{ $shardsPerZone }}
+  target_label: '__tmp_hash'
+  replacement: '$1'
+  action: 'hashmod'
+- source_labels: ['__tmp_hash']
+  separator: ;
+  regex: '{{ .operator.shardIndex | div $shardsPerZone }}'
+  replacement: '$1'
+  action: 'keep'
+```
+
+we would get the following output for `shard_index == 2`:
+
+```yaml
+- source_labels: [ '__meta_kubernetes_pod_label_topology_kubernetes_io_zone' ] 
+  separator: ;
+  regex: 'europe-west4-a'
+  replacement: $1
+  action: keep
+- source_labels: [ '__address__' ] 
+  separator: ;
+  modulus: 2
+  target_label: '__tmp_hash'
+  replacement: '$1'
+  action: 'hashmod'
+- source_labels: [ '__tmp_hash' ]
+  separator: ;
+  regex: '1'
+  replacement: '$1'
+  action: 'keep'
+```
+
+## Prometheus instance zone assignment
+
+To make sure that prometheus instances are deployed to the correct zone (of their
+assigned target), we need to generate a [node affinity](https://kubernetes.io/docs/concepts/scheduling-eviction/assign-pod-node/#node-affinity)
+or a [node selector](https://kubernetes.io/docs/tasks/configure-pod-container/assign-pods-nodes/#create-a-pod-that-gets-scheduled-to-your-chosen-node).
+
+As node selectors are simpler to manage, and node affinities might run into
+ordering issues when a user defines their own affinities, node selectors should
+be used.
+
+We use the same Helm templating notation as above to illustrate the mechanics.
+
+```yaml
+{{ $shardsPerZone := max 1 (.spec.shards | div (len .spec.shardingStrategy.topology.values)) }}
+spec:
+  nodeSelector:
+    '{{ .spec.shardingStrategy.topology.nodeLabel }}': '{{ index .spec.shardingStrategy.topology.values (.operator.shardIndex | div $shardsPerZone) }}'
+```
+
+Given this input:
+
+```yaml
+spec:
+  shards: 4
+  shardingStrategy:
+    mode: 'Topology'    
+    additionalRelabelConfig: [] 
+    topology:
+      nodeLabel: 'topology.kubernetes.io/zone'
+      sourceLabel: '__meta_kubernetes_pod_label_topology_kubernetes_io_zone'
+      values:
+        - europe-west4-a
+        - europe-west4-b
+```
+
+The following snippet would be generated for `shared_index == 2`:
+
+```yaml
+spec:
+  nodeSelector:
+    'topology.kubernetes.io/zone': 'europe-west4-a'
+```
+
+## Alternatives
+
+N/A
```

Comment:
An alternative could be to expose a way for users to customize the shard relabelings and let them manage all the settings manually (nodeSelector + relabelings) but it would be way more less user-friendly.

Reply:
added